tests/complex_contracts/nft/erc721.rs

```diff
@@ -0,0 +1,993 @@
+    );
+
+    let (_, _) = deploy(
+        &mut state,
+        "starknet_programs/ERC165.json",
+        &[],
+        &general_config,
+    );
+
+    let entry_points_by_type = state
+        .get_contract_class(&class_hash)
+        .unwrap()
+        .entry_points_by_type()
+        .clone();
+
+    let caller_address = Address(666.into());
+    let general_config = StarknetGeneralConfig::default();
+    let mut resources_manager = ExecutionResourcesManager::default();
+    let entry_point_type = EntryPointType::External;
+
+    let mut call_config = CallConfig {
+        state: &mut state,
+        caller_address: &caller_address,
+        address: &contract_address,
+        class_hash: &class_hash,
+        entry_points_by_type: &entry_points_by_type,
+        entry_point_type: &entry_point_type,
+        general_config: &general_config,
+        resources_manager: &mut resources_manager,
+    };
+
+    // data_len = 0 then there is no need to sent *felt for data
+    let calldata = [
+        Felt252::from(666),
+        Felt252::from(1000000),
+        Felt252::from(1),
+        Felt252::zero(),
+        Felt252::zero(),
+    ]
+    .to_vec();
+
+    // The contract will fail because the receiver address is not a IERC721Receiver contract.
+    assert!(execute_entry_point("safeTransferFrom", &calldata, &mut call_config).is_err());
+}
+
+#[test]
+fn erc721_calling_constructor_twice_should_fail_test() {
+    let general_config = StarknetGeneralConfig::default();
+    let mut state = CachedState::new(InMemoryStateReader::default(), Some(Default::default()));
+
+    let collection_name = Felt252::from_bytes_be("some-nft".as_bytes());
+    let collection_symbol = Felt252::from(555);
+    let to = Felt252::from(666);
+    let calldata = [collection_name, collection_symbol, to].to_vec();
+
+    let (contract_address, class_hash) = deploy(
+        &mut state,
+        "starknet_programs/ERC721.json",
+        &calldata,
+        &general_config,
+    );
+
+    let caller_address = Address(666.into());
+    let general_config = StarknetGeneralConfig::default();
+    let mut resources_manager = ExecutionResourcesManager::default();
+    let entry_point_type = EntryPointType::Constructor;
+
+    let entry_points_by_type = state
+        .get_contract_class(&class_hash)
+        .unwrap()
+        .entry_points_by_type()
+        .clone();
+
+    let mut call_config = CallConfig {
+        state: &mut state,
+        caller_address: &caller_address,
+        address: &contract_address,
+        class_hash: &class_hash,
+        entry_points_by_type: &entry_points_by_type,
+        entry_point_type: &entry_point_type,
+        general_config: &general_config,
+        resources_manager: &mut resources_manager,
+    };
+
+    assert!(execute_entry_point("constructor", &calldata, &mut call_config).is_err());
+}
+
+//Should panic is necessary because the constructor will fail
+//deploy() will try to unwrap the result of the constructor
```

Comment:
We could change the _utils_ functions to return results. That way they would be more flexible and allow us to check the exact errors returned when needed.

Reply:
Great idea! Refactor done